### PR TITLE
[PB-13] Port benchmarking code to develop

### PR DESCRIPTION
**Pull request by 365andreas**

### CODE REVIEW

File: modules/network-wide.nix [hunk 1]
```diff
@@ -2,7 +2,8 @@
 
 with import ./../lib.nix;
 {
-  config.deployment = {
+  config = mkIf (false) {
```
Comment (CodiePP) [inline modules/network-wide.nix]: can we delete this?

File: destroy-cluster.nix [hunk 2]
```diff
@@ -0,0 +1,11 @@
+with import <nixpkgs> {};
+writeScriptBin "destroy-cluster.sh" ''
+  #!/usr/bin/env bash
+
+  set -e        # exit on error
+  set -o xtrace # print commands
+
+  IO=$(nix-build -A iohk-ops)/bin/iohk-ops
```
Comment (cleverca22) [inline destroy-cluster.nix]: this could be improved to just refer to io directly

File: modules/cardano-service.nix [hunk 3]
```diff
@@ -176,13 +176,14 @@ in {
       nodeMinute = mod (cfg.nodeIndex * 4) 60;
     in {
       script = ''
-        /run/current-system/sw/bin/systemctl restart cardano-node
+        echo /run/current-system/sw/bin/systemctl restart cardano-node
```
[the remark this answers is not in the document]
Comment (cleverca22) [reply modules/cardano-service.nix]: yeah, that should be fairly simple to add

File: modules/cardano-service.nix [hunk 4]
```diff
@@ -49,6 +48,7 @@ let
     "--configuration-key ${config.deployment.arguments.configurationKey}"
     "--topology ${cfg.topologyYaml}"
     "--node-id ${params.name}"
+    ("--policies " + (if ("c" == builtins.substring 0 1 "${params.name}") then "${./../policy_core.yaml}" else "${./../policy_relay.yaml}"))
```
Comment (cleverca22) [inline modules/cardano-service.nix]: the node type may also be able to set this correctly

File: modules/network-wide.nix (hunk 1 above)
Comment (cleverca22) [reply modules/network-wide.nix]: probably

File: staging-testnet.yaml [hunk 5]
```diff
@@ -1,36 +1,30 @@
 args:
   configurationKey:
     tag: NixStr
-    contents: mainnet_dryrun_full
+    contents: bench
```
Comment (cleverca22) [inline staging-testnet.yaml]: this should go into a new file, and the original content restored

File: modules/cardano-service.nix (hunk 4 above)
Comment (365andreas) [reply modules/cardano-service.nix]: https://github.com/input-output-hk/iohk-ops/blob/50df339b40b5f75d796eb2e22ead02cbad4cd5c5/modules/cardano-service.nix#L51

Is the above ok?

File: destroy-cluster.nix (hunk 2 above)
Comment (365andreas) [reply destroy-cluster.nix]: @cleverca22 
Replacing `$IO` with `io` does not work saying `io : command not found`.

Script is being called inside a nix' shell.
(` $(nix-build destroy-cluster.nix)/bin/destroy-cluster.sh`)

File: staging-testnet.yaml (hunk 5 above)
Comment (365andreas) [reply staging-testnet.yaml]: Done.

File: modules/cardano-service.nix (hunk 3 above)
[the remark this answers is not in the document]
Comment (365andreas) [reply modules/cardano-service.nix]: https://github.com/input-output-hk/iohk-ops/blob/e57ac51a37eeab53567efce7a76e6b04eedf79ee/modules/cardano-service.nix#L175-L181

Is the above ok?
How can I check it?

File: modules/cardano-service.nix [hunk 6]
```diff
@@ -214,5 +215,19 @@ in {
         Type = "notify";
       };
     };
+
+    systemd.services.cardano-node-recorder = {
```
[the remark this answers is not in the document]
Comment (365andreas) [reply modules/cardano-service.nix]: https://github.com/input-output-hk/iohk-ops/blob/e57ac51a37eeab53567efce7a76e6b04eedf79ee/modules/cardano-service.nix#L217-L220

File: deployments/edgenodesv2.nix [hunk 7]
```diff
@@ -1,32 +1,47 @@
...
-      topologyFile = "${topologyFile}";
+  mkService = index: {
+    users.users."cardano-node-${toString index}" = {
+      isNormalUser = true;
+    };
+    systemd.services."cardano-node-${toString index}" = {
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig = {
+        User = "cardano-node-${toString index}";
+        WorkingDirectory = "/home/cardano-node-${toString index}";
```
Comment (cleverca22) [inline deployments/edgenodesv2.nix]: can you revert all changes to this file relative to develop?, this has already been merged in another PR and these changes are undoing some improvements

File: modules/cardano-edgenode.nix [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+let
+  instancesPerNode = 5;
+  params = {
+    name = "wallet";
+  };
+  mkNode = publicIP: nodes: index: {
+    name = "instance${toString index}";
+    value = {
+      autoStart = true;
+      privateNetwork = false;
```
Comment (cleverca22) [inline modules/cardano-edgenode.nix]: and just delete this file, its been replaced by `cardano-benchmark.nix`

File: modules/cardano-service.nix [hunk 9]
```diff
@@ -174,14 +176,15 @@ in {
...
       # Reboot cardano-node every 36h (except Mon->Tue gap which is 24h)
-      startAt = [
-        "Tue,Fri,Mon 13:${toString nodeMinute}"
-        "Thu,Sun     01:${toString nodeMinute}"
-      ];
+      #startAt = [
+      #  "Tue,Fri,Mon 13:${toString nodeMinute}"
+      #  "Thu,Sun     01:${toString nodeMinute}"
+      #];
+      startAt = lib.mkForce [];
```
Comment (cleverca22) [inline modules/cardano-service.nix]: and revert the change to `startAt`, the `script` change can stay

File: scripts/ci.sh [hunk 10]
```diff
@@ -108,6 +110,13 @@ ${IOHK_OPS} "${GENERAL_OPTIONS[@]}" --config 'test-infra.yaml'  create deploy --
 banner 'Staging infra evaluated'
 fi
 
+if test -n "${WITH_BENCHMARK}"; then
+CLEANUP_DEPLS="${CLEANUP_DEPLS} test-bench"
+${IOHK_OPS}               new  --config 'test-bench.yaml'   --environment benchmark    ${COMMON_OPTIONS} 'test-stag'    ${CARDANO_COMPONENTS}
+${IOHK_OPS} ${GENERAL_OPTIONS} --config 'test-bench.yaml'   create deploy --dry-run
+banner 'Benchmark env evaluated'
```
Comment (cleverca22) [inline scripts/ci.sh]: i think the topology file might be missing in this area, causing CI to fail

File: deployments/edgenodesv2.nix (hunk 7 above)
Comment (365andreas) [reply deployments/edgenodesv2.nix]: Done.

File: modules/cardano-edgenode.nix (hunk 8 above)
Comment (365andreas) [reply modules/cardano-edgenode.nix]: Done.

File: modules/cardano-service.nix (hunk 9 above)
Comment (365andreas) [reply modules/cardano-service.nix]: Done.

File: scripts/ci.sh (hunk 10 above)
Comment (365andreas) [reply scripts/ci.sh]: Done.

**Comment by 365andreas** [comment]
Also cherry-picked: 
https://github.com/input-output-hk/iohk-ops/pull/338/commits/4309a0f6b12c0a18c2bdc6a70c79e7d603d5949c

**Comment by jmitchell** [comment]
Shellcheck isn't happy about the new scripts. In this case you can add an [ignore comment](https://github.com/koalaman/shellcheck/wiki/Ignore) if you'd like.

**Comment by 365andreas** [comment]
@jmitchell Fixed it.

**Comment by CodiePP** [comment]
we would like to have this merged, so we can continue.
@cleverca22 : is there still some issue open? it looks like @365andreas addressed all your change requests.
